### PR TITLE
Add entity buffer and access check to taxonomy tree data producer.

### PR DESCRIPTION
**Pull request by rthideaway**
Adds missing access check for taxonomy tree data producer. Use buffer for loading terms.

### CODE REVIEW

File: src/Plugin/GraphQL/DataProducer/Taxonomy/TaxonomyLoadTree.php [hunk 1]
```diff
@@ -110,7 +149,43 @@ public function resolve(string $vid, int $parent = 0, ?int $max_depth = NULL, Re
       ->getStorage('taxonomy_term')
       ->loadTree($vid, $parent, $max_depth);
 
-    return $terms;
+    $term_ids = array_column($terms, 'tid');
+    $resolver = $this->entityBuffer->add('taxonomy_term', $term_ids);
+
+    return new Deferred(function () use ($type, $ids, $language, $resolver, $context, $access, $accessUser, $accessOperation) {
```
Comment (klausi) [inline src/Plugin/GraphQL/DataProducer/Taxonomy/TaxonomyLoadTree.php]: $language is undefined? Same for $ids? Please check all variables if they actually exist here.

**Review by klausi — COMMENTED** [review]
There is a todo comment that should be removed now in resolve().

I'm not sure if we should load full term objects since that can be slow. I think that was the reason why we just started with the fast term tree loading.

In jobiqo we just allow access to all taxonomies, so access protection is not a concern for us for now.

Translating terms might be a use case? It does not seem to be working because there is no language parameter?

File: src/Plugin/GraphQL/DataProducer/Taxonomy/TaxonomyLoadTree.php (hunk 1 above)
Comment (rthideaway) [reply src/Plugin/GraphQL/DataProducer/Taxonomy/TaxonomyLoadTree.php]: Fixed language usage, removed ids and removed todo. Tested and it worked.

As we agreed, for our use case I (re)introduce secondary data producer which just loads stdClass objects without language and access checks

File: src/Plugin/GraphQL/DataProducer/Taxonomy/TaxonomyLoadTree.php [hunk 2]
```diff
@@ -95,22 +132,66 @@ public function __construct(
...
+        if (isset($language) && $language !== $entities[$id]->language()->getId() && $entities[$id] instanceof TranslatableInterface) {
+          $entities[$id] = $entities[$id]->getTranslation($language);
+          $entities[$id]->addCacheContexts(["static:language:{$language}"]);
+        }
+
+        if ($access) {
+          /* @var $accessResult \Drupal\Core\Access\AccessResultInterface */
+          $accessResult = $entity->access($accessOperation, $accessUser, TRUE);
+          $context->addCacheableDependency($accessResult);
+          if ($accessResult->isForbidden()) {
```
Comment (fubhy) [inline src/Plugin/GraphQL/DataProducer/Taxonomy/TaxonomyLoadTree.php]: Has to be ->isAllowed() as @klausi just found in the other issue. This needs to b e fixed everywhere in our current code base. Maybe you can open a PR that just replaces ALL occurrences of `isForbidden` with `isAllowed` (and negating the condition) for the entire code base?

**Review by fubhy — CHANGES_REQUESTED** [review]

File: src/Plugin/GraphQL/DataProducer/Taxonomy/TaxonomyLoadTree.php (hunk 2 above)
Comment (rthideaway) [reply src/Plugin/GraphQL/DataProducer/Taxonomy/TaxonomyLoadTree.php]: Adjusted. Should be ready for review

**Comment by fubhy** [comment]
Thanks!

**Pull request merged**